src/CSnakes.Runtime/Python/Pack.cs

```diff
@@ -10,46 +13,155 @@ namespace CSnakes.Runtime.Python;
+            spillLength > 0
+            ? new PyObjectMarshaller.ManagedToUnmanagedIn[spillLength]
+            : null;
+
+        var uninitializedMarshallers = MemoryMarshal.CreateSpan(ref Unsafe.As<ArrayOf8<PyObjectMarshaller.ManagedToUnmanagedIn>, PyObjectMarshaller.ManagedToUnmanagedIn>(ref initialMarshallers), stackSpillThreshold);
+        var uninitializedHandles = initialHandles;
+
+        // The following loop initializes the marshallers and handles for each
+        // item in the input span. It is assumed that no exceptions are thrown
+        // during this loop. The marshallers are freed in the finally block of
+        // the actual list/tuple initialization.
+
+        foreach (var item in items)
         {
-            var handles = items.Length < 18 // .NET tuples are max 17 items. This is a performance optimization.
-                ? stackalloc IntPtr[items.Length]
-                : new IntPtr[items.Length];
+            PyObjectMarshaller.ManagedToUnmanagedIn m = default;
+            m.FromManaged(item);
 
-            for (int i = 0; i < items.Length; i++)
+            if (uninitializedMarshallers.IsEmpty)
             {
-                SafeHandleMarshaller<PyObject>.ManagedToUnmanagedIn m = default;
-                m.FromManaged(items[i]);
-                marshallers.Add(m);
-                handles[i] = m.ToUnmanaged();
+                Debug.Assert(spilledMarshallers is not null);
+                uninitializedMarshallers = spilledMarshallers;
             }
-            return PyObject.Create(CPythonAPI.PackTuple(handles));
-        }
-        finally
-        {
-            foreach (var m in marshallers)
+
+            uninitializedMarshallers[0] = m;
+            uninitializedMarshallers = uninitializedMarshallers[1..];
+
+            if (uninitializedHandles.IsEmpty)
             {
-                m.Free();
+                Debug.Assert(spilledHandles is not null);
+                uninitializedHandles = spilledHandles;
             }
+
+            uninitializedHandles[0] = m.ToUnmanaged();
+            uninitializedHandles = uninitializedHandles[1..];
         }
-    }
 
-    internal static PyObject CreateList(Span<PyObject> items)
-    {
-        PyObject pyList = PyObject.Create(CPythonAPI.PyList_New(0)); // TODO: preallocate based on items.Length and use PyList_SetItem
+        nint obj = 0;
 
-        foreach (var item in items)
+        try
         {
-            int result = CPythonAPI.PyList_Append(pyList, item);
-            if (result == -1)
+            obj = TBuilder.New(items.Length);
+            SetItems(spilledHandles, SetItems(initialHandles, 0));
+
+            return obj;
+
+            int SetItems(Span<nint> handles, int i)
             {
-                throw PyObject.ThrowPythonExceptionAsClrException();
+                foreach (var handle in handles)
+                {
+                    int result = TBuilder.SetItemRaw(obj, i++, handle);
```

Comment:
I still see a capture here. Am I commenting prior to you pushing up changes? If so, apologies.

Reply:
Ah, sorry, my bad. I just realised the change was still sitting locally and made the wrong reference in [my earlier comment](https://github.com/tonybaloney/CSnakes/pull/206#discussion_r1774089055). Just committed and pushed as 20a59e145ada5a5250d1e40414552fa15f89b8f8. As a note to self, don't work late and tired. 😅